libevmasm/Assembly.cpp

```diff
@@ -884,175 +918,292 @@ LinkerObject const& Assembly::assemble() const
+		_dest.resize(_dest.size() + _size);
+		setBigEndian(_dest, _dest.size() - _size, _size, _value);
+	};
+	static auto appendBigEndianUint16 = [](bytes& _dest, auto _value) {
+		static_assert(!std::numeric_limits<decltype(_value)>::is_signed, "only unsigned types or bigint supported");
+		assertThrow(_value <= 0xFFFF, AssemblyException, "");
+		appendBigEndian(_dest, 2, static_cast<size_t>(_value));
+	};
+	std::vector<size_t> codeSectionSizeOffsets;
+	auto setCodeSectionSize = [&](size_t _section, size_t _size) {
+		if (eof)
+			toBigEndian(_size, bytesRef(ret.bytecode.data() + codeSectionSizeOffsets.at(_section), 2));
+	};
+	std::optional<size_t> dataSectionSizeOffset;
+	auto setDataSectionSize = [&](size_t _size) {
+		if (eof)
+		{
+			assertThrow(dataSectionSizeOffset.has_value(), AssemblyException, "");
+			assertThrow(_size <= 0xFFFF, AssemblyException, "Invalid data section size.");
+			toBigEndian(_size, bytesRef(ret.bytecode.data() + *dataSectionSizeOffset, 2));
+		}
+	};
+
+	size_t startOfContainerSectionHeader = 0;
+
+	// Insert EOF1 header.
+	if (eof)
+	{
+		ret.bytecode.push_back(0xef);
+		ret.bytecode.push_back(0x00);
+		ret.bytecode.push_back(0x01);										// version 1
+
+		ret.bytecode.push_back(0x01);										// kind=type
+		appendBigEndianUint16(ret.bytecode, m_codeSections.size() * 4u);	// length of type section
+
+		ret.bytecode.push_back(0x02);										// kind=code
+		appendBigEndianUint16(ret.bytecode, m_codeSections.size());			// placeholder for number of code sections
+
+		for (auto const& codeSection: m_codeSections)
+		{
+			(void) codeSection;
+			codeSectionSizeOffsets.emplace_back(ret.bytecode.size());
+			appendBigEndianUint16(ret.bytecode, 0u);						// placeholder for length of code
+		}
+
+		startOfContainerSectionHeader = ret.bytecode.size();
+
+		ret.bytecode.push_back(0x04);										// kind=data
+		dataSectionSizeOffset = ret.bytecode.size();
+		appendBigEndianUint16(ret.bytecode, 0u);							// length of data
+
+		ret.bytecode.push_back(0x00);										// terminator
+
+		for (auto const& codeSection: m_codeSections)
+		{
+			ret.bytecode.push_back(codeSection.inputs);
+			ret.bytecode.push_back(codeSection.outputs);
+			appendBigEndianUint16(ret.bytecode, 0xFFFFu);	// TODO: Add stack height calculation
+		}
+	}
+
+	unsigned headerSize = eof ? static_cast<unsigned>(ret.bytecode.size()) : 0;
 	unsigned bytesRequiredForCode = codeSize(static_cast<unsigned>(subTagSize));
 	m_tagPositionsInBytecode = std::vector<size_t>(m_usedTags, std::numeric_limits<size_t>::max());
 	std::map<size_t, std::pair<size_t, size_t>> tagRef;
 	std::multimap<h256, unsigned> dataRef;
 	std::multimap<size_t, size_t> subRef;
 	std::vector<unsigned> sizeRef; ///< Pointers to code locations where the size of the program is inserted
-	unsigned bytesPerTag = numberEncodingSize(bytesRequiredForCode);
+	unsigned bytesPerTag = numberEncodingSize(headerSize + bytesRequiredForCode + bytesRequiredForDataUpperBound);
```

Comment:
It seems to me that this values will now be different, even for non-EOF code. It now includes the size of `m_auxiliaryData` while before it didn't. On the other hand the extra byte we add to `bytesRequiredForCode` below is not included here.

Not yet sure if this can break anything. Adding too much will probably just make us waste some gas?